src/v/lsm/io/disk_persistence.cc

```diff
@@ -0,0 +1,352 @@
+
+#include "base/units.h"
+#include "lsm/core/exceptions.h"
+#include "lsm/io/persistence.h"
+#include "utils/uuid.h"
+
+#include <seastar/core/fstream.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <exception>
+#include <system_error>
+
+namespace lsm::io {
+
+namespace {
+
+class disk_seq_file_reader : public sequential_file_reader {
+public:
+    disk_seq_file_reader(
+      std::filesystem::path path, ss::input_stream<char> stream)
+      : _path(std::move(path))
+      , _stream(std::move(stream)) {}
+
+    ss::future<iobuf> read(size_t n) override {
+        iobuf buf;
+        try {
+            while (n > 0) {
+                auto tmp_buf = co_await _stream.read_up_to(
+                  std::min(n, 128_KiB));
+                if (tmp_buf.empty()) {
+                    break;
+                }
+                n -= tmp_buf.size();
+                buf.append(
+                  std::make_unique<iobuf::fragment>(std::move(tmp_buf)));
+            }
+        } catch (const std::system_error& err) {
+            throw io_error_exception(
+              err.code(), "io error seq reading: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error seq reading: {}", std::current_exception());
+        }
+        co_return buf;
+    }
+    ss::future<> skip(size_t n) override {
+        try {
+            co_await _stream.skip(n);
+        } catch (const std::system_error& err) {
+            throw io_error_exception(
+              err.code(), "io error seq skipping: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error seq skipping: {}", std::current_exception());
+        }
+    }
+    ss::future<> close() override {
+        try {
+            co_await _stream.close();
+        } catch (const std::system_error& err) {
+            throw io_error_exception(err.code(), "io error closing: {}", err);
+        } catch (...) {
+            throw io_error_exception(
+              "io error closing: {}", std::current_exception());
+        }
+    }
+
+    fmt::iterator format_to(fmt::iterator it) const override {
+        return fmt::format_to(it, "{{path={}}}", _path);
+    }
+
+private:
+    std::filesystem::path _path;
+    ss::input_stream<char> _stream;
+};
+
+class disk_file_reader : public random_access_file_reader {
+public:
+    disk_file_reader(std::filesystem::path path, ss::file file)
+      : _path(std::move(path))
+      , _file(std::move(file)) {}
+
+    ss::future<ioarray> read(size_t offset, size_t n) override {
+        size_t memory_alignment = _file.memory_dma_alignment();
+        size_t disk_alignment = _file.disk_read_dma_alignment();
+        size_t adjusted_offset = ss::align_down(offset, disk_alignment);
+        size_t offset_delta = offset - adjusted_offset;
+        auto array = ioarray::aligned(
+          memory_alignment, ss::align_up(n + offset_delta, disk_alignment));
```

Comment:
hmm, i think this may address my previous question about trimming and how it relates to I/O size.